src/peerDiscovery.cpp

```diff
@@ -0,0 +1,378 @@
+
+    uint opt = 1;
+
+    if (setsockopt(m_sockfd, 
+                   SOL_SOCKET, 
+                   SO_REUSEADDR|SO_REUSEPORT, 
+                   (char*) &opt, 
+                   sizeof(opt)) < 0
+    ){
+       throw std::runtime_error(std::string("Reusing ADDR failed_initialize_socketopt") + strerror(errno));
+    }
+
+}
+
+void
+PeerDiscovery::initialize_sockaddr_Listener()
+{   
+    bzero(&m_sockaddr,sizeof(m_sockaddr));
+    m_sockaddr.setFamily(m_domain);
+    m_sockaddr.setPort(m_port);
+    m_sockaddr.setAny();
+
+    // bind to receive address
+    if (bind(m_sockfd, 
+             m_sockaddr.get(), 
+             m_sockaddr.getLength()) < 0
+    ){
+
+        throw std::runtime_error(std::string("bind_mcast_join") + strerror(errno));
+
+    }
+
+}
+
+void
+PeerDiscovery::initialize_sockaddr_Sender(){
+
+    bzero(&m_sockaddr,sizeof(m_sockaddr));
+    m_sockaddr.setFamily(m_domain);
+    m_sockaddr.setPort(m_port);
+    m_sockaddr.setAddress(multicast_address_ipv4,multicast_address_ipv6);
+
+}
+
+void
+PeerDiscovery::mcast_join(){
+
+    switch (m_domain)
+    {
+        case AF_INET:
+
+            struct ip_mreq config_ipv4;
+            // config the listener to be interested in joining in the multicast group
+            config_ipv4.imr_multiaddr.s_addr = inet_addr(multicast_address_ipv4);
+            config_ipv4.imr_interface.s_addr = htonl(INADDR_ANY);
+
+            if (setsockopt(m_sockfd, 
+                           IPPROTO_IP, 
+                           IP_ADD_MEMBERSHIP, 
+                           (char*)&config_ipv4, 
+                           sizeof(config_ipv4)) < 0
+            ){
+                throw std::runtime_error(std::string("Setsockopt_mcast_join error:") + strerror(errno));
+            }
+
+            break;
+    
+        case AF_INET6:
+
+            struct ipv6_mreq config_ipv6;
+            config_ipv6.ipv6mr_interface = 0;
+            inet_pton(AF_INET6, multicast_address_ipv6, &config_ipv6.ipv6mr_multiaddr);
+            if (setsockopt(m_sockfd, 
+                           IPPROTO_IPV6, 
+                           IPV6_ADD_MEMBERSHIP, 
+                           &config_ipv6, 
+                           sizeof(config_ipv6)) < 0
+            ){
+                throw std::runtime_error(std::string("setsockopt_mcast_join") + strerror(errno));
+            }
+
+            break;
+
+    }       
+
+}
+
+void
+PeerDiscovery::m_sendto(uint8_t *buf,size_t &buf_size){
+
+    int nbytes = sendto(
+        m_sockfd,
+        buf,
+        buf_size,
+        0,
+        m_sockaddr.get(),
+        m_sockaddr.getLength()
+    );
+    if (nbytes < 0) {
+
+        throw std::runtime_error(std::string("sendto_m_sendto") + strerror(errno));
+
+    }
+
+}
+
+void
+PeerDiscovery::m_recvfrom(uint8_t *buf,size_t &buf_size)
+{
+    socklen_t sa_len = sizeof(m_sockaddr);
+    int nbytes = recvfrom(
+        m_sockfd,
+        buf,
+        buf_size,
+        0,
+        m_sockaddr.get(),
```

Comment:
could cause memory corruption since we don't know the sender address length. declare sockaddr_storage on the stack and return it.